[UI] Bootstrap the layout of dashboard page

ChengYanJin opened the pull request:
**Component**: ui

**Context**: 
Bootstrap the layout of the dashboard page, we called it overview page previously.

**Summary**:
- Responsiveness 
  CSS Grid
- Add a feature gate to control access to the dashboard page.
   add `flags: ["dashboard"]` in `config.json` or `ConfigMap` will enable the access. By default, no access.
- We still serve the alert page as the main page, meaning`https://{ip}:8443/` takes us to the alert page.

**Acceptance criteria**: 
Mac 13 inch screen:
![image](https://user-images.githubusercontent.com/18453133/108687237-c1f88a80-74f6-11eb-891d-d1cbf982a939.png)

1024*768
![image](https://user-images.githubusercontent.com/18453133/108689167-09801600-74f9-11eb-8e7b-beb7cf6f3165.png)
![image](https://user-images.githubusercontent.com/18453133/108689965-0f2a2b80-74fa-11eb-86fd-94859e430e10.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3094 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/dashboard: Add dashboard page with layout

Bootstrap the layout by CSS Grid

Refs: #3094

ChengYanJin committed:
ui/translations: Add translations for the dashboard

Refs: #3094

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/3094-bootstrap-ui-overview-page` into
`development/2.8`
* [w/2.9/feature/3094-bootstrap-ui-overview-page](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/feature/3094-bootstrap-ui-overview-page) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

JBWatenbergScality commented on ui/src/containers/DashboardPage.js:
```diff
@@ -0,0 +1,116 @@
+  font-size: ${fontSize.large};
+`;
+
+const DashboardPage = (props: Object) => {
```
For now it takes no props, use `{}` type

JBWatenbergScality commented on ui/src/containers/DashboardPage.js:
```diff
@@ -0,0 +1,116 @@
+  width: 50%;
+  background-color: ${(props) => props.theme.brand.primary};
+  margin-left: 8px;
+`;
```
I would recommend to try to abstract a bit those cards to a single component and define their heights and other specificities as a style prop. That would simplify switching to real content because in a real content world the height of the cards will be computed according to their contents.

JBWatenbergScality commented on ui/src/ducks/config.js:
```diff
@@ -210,6 +218,9 @@ export function* fetchConfig(): Generator<Effect, void, Result<Config>> {
   }
+  if (!result.error && result.flags) {
+    yield put(setFeatureGateAction(result.flags));
+  }
```
flags are already pushed to redux store here : https://github.com/scality/metalk8s/pull/3142/files#diff-2330a8a0385ac1477f3c2323dd6207b351dbc65f93c33c310fddab466f004b49R202 I don't think that we need a new specific action to retrieve them

ChengYanJin replied:
Good point! Removed!

JBWatenbergScality commented on ui/src/translations/fr.json:
```diff
@@ -185,5 +185,6 @@
   "show_cluster_avg": "Montrer la moyenne du Cluster",
-  "cluster_avg": "Cluster moy"
+  "cluster_avg": "Cluster moy",
+  "dashboard": "Dashboard"
```
Maybe we should translate this ?

```suggestion
  "dashboard": "Tableau de bord"
```

ChengYanJin replied:
After discussing with @Cuervino,  we decided not to translate the "dashboard" for the following reason.
1. In french, we also use "dashboard" in the professional environment.
2. "Tableau de bord" would be too long as a sidebar item.

bert-e commented:
# History mismatch



Merge commit #840e993bf50807ff2cac32a29f58f7ec1a205f90 on the integration branch
`w/2.9/feature/3094-bootstrap-ui-overview-page` is merging a branch which is neither the current
branch `feature/3094-bootstrap-ui-overview-page` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/3094-bootstrap-ui-overview-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin committed:
ui/nodes: Add color field to remove the warning in the console

Refs: #3094

gdemonet reviewed (commented):
Unpopular opinion: this layout is a grid, using Flexbox and `calc` with a combination of `vh`, `%`, and `px`, makes it impossible to understand at first read, hence impossible to maintain.

CSS Grid was built to solve this specific use case:

```css
.dashboard {
  display: grid;
  gap: 10px;
  grid-template:
      "health    health   network metrics" 160px
      "inventory services network metrics" 160px
      "alerts    services network metrics" auto
    / 1fr        1fr      1fr     1fr;
}

@media (max-width: 1024px) {
  .dashboard {
    grid-template:
        "health    health"   160px
        "inventory services" 160px
        "alerts    services" minmax(160px, auto)
        "network   metrics"  auto
      / 1fr        1fr;
  }
}
```
  

(see it in action: https://cdpn.io/gdemonet/debug/GRNMORw/NjrYzGWojBRA)

JBWatenbergScality commented:
> Unpopular opinion: this layout is a grid, using Flexbox and `calc` with a combination of `vh`, `%`, and `px`, makes it impossible to understand at first read, hence impossible to maintain.
> 
> CSS Grid was built to solve this specific use case:
> 
> ```css
> .dashboard {
>   display: grid;
>   gap: 10px;
>   grid-template:
>       "health    health   network metrics" 160px
>       "inventory services network metrics" 160px
>       "alerts    services network metrics" auto
>     / 1fr        1fr      1fr     1fr;
> }
> 
> @media (max-width: 1024px) {
>   .dashboard {
>     grid-template:
>         "health    health"   160px
>         "inventory services" 160px
>         "alerts    services" minmax(160px, auto)
>         "network   metrics"  auto
>       / 1fr        1fr;
>   }
> }
> ```
> 
> (see it in action: https://cdpn.io/gdemonet/debug/GRNMORw/NjrYzGWojBRA)

This implementation proposal would solves https://github.com/scality/metalk8s/pull/3142#discussion_r580114702

ChengYanJin committed:
ui/dashboard: Add dashboard page to layout with feature gate

Refs: #3094

ChengYanJin committed:
ui/config: Add flags to config.json as feature gate

We implement feature gate for the dashboard page so that we can
merge the uncomplete work and the users won't detect anything unless
they add `flags:['dashboard']` to configMap/config.json.

Refs: #3094

JBWatenbergScality commented on ui/src/containers/Layout.js:
```diff
@@ -56,8 +61,22 @@ const Layout = (props) => {
+        label: intl.translate('dashboard'),
         icon: <i className="fas fa-desktop" />,
+        onClick: () => {
+          history.push('/dashboard');
```
I think dashboard should be served on `/` and alerts should be served on `/alerts`. Also alerts screen should be removed from this side bar as it will be added in the navbar soon

ChengYanJin replied:
Absolutely!
When the navbar is ready, we can consider moving the alert page to there with some design changes too. 
I will open a ticket for this once we have the design from @Cuervino.

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# History mismatch



Merge commit #840e993bf50807ff2cac32a29f58f7ec1a205f90 on the integration branch
`w/2.9/feature/3094-bootstrap-ui-overview-page` is merging a branch which is neither the current
branch `feature/3094-bootstrap-ui-overview-page` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/3094-bootstrap-ui-overview-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/3094-bootstrap-ui-overview-page` into
`development/2.8`
* [w/2.9/feature/3094-bootstrap-ui-overview-page](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/feature/3094-bootstrap-ui-overview-page) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

ChengYanJin committed:
ui/cypress: cy.visit() to directly access the page

Since we already implemented interaction/integreation tests to test the
navigation of the sidebar. We don't need to duplicate the test here.
So we should directly access the page by `cy.visit()`

Refs: #3094

bert-e commented:
# History mismatch



Merge commit #ebdf2313e47741e7c393ebb8236e199aee8442f2 on the integration branch
`w/2.9/feature/3094-bootstrap-ui-overview-page` is merging a branch which is neither the current
branch `feature/3094-bootstrap-ui-overview-page` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/3094-bootstrap-ui-overview-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# History mismatch



Merge commit #ebdf2313e47741e7c393ebb8236e199aee8442f2 on the integration branch
`w/2.9/feature/3094-bootstrap-ui-overview-page` is merging a branch which is neither the current
branch `feature/3094-bootstrap-ui-overview-page` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/3094-bootstrap-ui-overview-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/3094-bootstrap-ui-overview-page` into
`development/2.8`
* [w/2.9/feature/3094-bootstrap-ui-overview-page](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/feature/3094-bootstrap-ui-overview-page) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

bert-e commented:
# History mismatch



Merge commit #cdd5b54926dddc6abcf655d1b78d7d3fcbb6d51a on the integration branch
`w/2.9/feature/3094-bootstrap-ui-overview-page` is merging a branch which is neither the current
branch `feature/3094-bootstrap-ui-overview-page` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/3094-bootstrap-ui-overview-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/3094-bootstrap-ui-overview-page` into
`development/2.8`
* [w/2.9/feature/3094-bootstrap-ui-overview-page](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/feature/3094-bootstrap-ui-overview-page) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye chengyanjin.
